### PR TITLE
GenericClient POST paging support

### PR DESCRIPTION
**Pull request by volodymyr-korzh**
In order to support HTTP POST method for paging, enhanced GenericClient.GetPageInternal class to execute both GET or POST paging requests.

### CODE REVIEW

File: hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java [hunk 1]
```diff
@@ -61,12 +62,12 @@ public class RestServerR4Helper extends BaseRestServerHelper implements BeforeEa
 	protected final MyRestfulServer myRestServer;
 
 	public RestServerR4Helper() {
-		this(false, false);
```
Comment (epeartree) [inline hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java:64]: not required.  please remove.

File: hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java [hunk 2]
```diff
@@ -61,12 +62,12 @@ public class RestServerR4Helper extends BaseRestServerHelper implements BeforeEa
...
+		this(false, false, false);
 	}
 
-	private RestServerR4Helper(boolean theInitialize, boolean theTransactionLatchEnabled) {
+	private RestServerR4Helper(boolean theInitialize, boolean theTransactionLatchEnabled, boolean theExecutePostPagingRequestsEnabled) {
```
Comment (epeartree) [inline hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java:68]: same here, no need for that. u have a setting method below and that's enough.  GET is the norm and if someone will use POST, they can use the setter.

File: hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java [hunk 3]
```diff
@@ -268,6 +269,10 @@ public void enableTransactionLatch(boolean theTransactionLatchEnabled) {
 		myRestServer.setTransactionLatchEnabled(theTransactionLatchEnabled);
 	}
 
+	public void enableExecutePostPagingRequests(boolean theExecutePostPagingRequestsEnabled) {
```
Comment (epeartree) [inline hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java:272]: rename to setHttpMethodForPagingRequest(PageStyle thePageStyle).

File: hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java [hunk 4]
```diff
@@ -281,9 +286,12 @@ private static class MyRestfulServer extends RestfulServer {
 
 		private final boolean myInitialTransactionLatchEnabled;
 
-		public MyRestfulServer(FhirContext theFhirContext, boolean theInitialTransactionLatchEnabled) {
+		private boolean myExecutePostPagingRequestsEnabled;
```
Comment (epeartree) [inline hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java:289]: private PageStyle myPageStyle = PageStyle.GET;

File: hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java [hunk 5]
```diff
@@ -281,9 +286,12 @@ private static class MyRestfulServer extends RestfulServer {
...
+
+		public MyRestfulServer(FhirContext theFhirContext, boolean theInitialTransactionLatchEnabled, boolean theExecutePostPagingRequestsEnabled) {
 			super(theFhirContext);
 			myInitialTransactionLatchEnabled = theInitialTransactionLatchEnabled;
+			myExecutePostPagingRequestsEnabled = theExecutePostPagingRequestsEnabled;
```
Comment (epeartree) [inline hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java:294]: the above is not needed.

File: hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java [hunk 6]
```diff
@@ -298,7 +306,21 @@ protected boolean isTransactionLatchEnabled() {
...
 		}
-		
+
+		public void setExecutePostPagingRequestsEnabled(boolean theExecutePostPagingRequestsEnabled) {
+			myExecutePostPagingRequestsEnabled = theExecutePostPagingRequestsEnabled;
+		}
+
+		@Override
+		protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+			// emulate a GET request invocation in the case of a POST paging request, as POST paging is not supported by FHIR RestServer
```
Comment (epeartree) [inline hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java:316]: no need for the if/else if u use myPageStyle.

**Review by epeartree — CHANGES_REQUESTED** [review]

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/rest/api/PageStyleEnum.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package ca.uhn.fhir.rest.api;
+
+/**
+ * Enumerated type to represent allowable syntax for a paging request.
+ */
```
[the remark this answers is not in the document]
Comment (volodymyr-korzh) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/rest/api/PageStyleEnum.java:24]: I choose `PageStyleEnum` class name in accordance to similar logic related to search request invocation, see `SearchStyleEnum`. But let me know if you think that `PageTypeEnum` is more appropriate class name

File: hapi-fhir-client/src/main/java/ca/uhn/fhir/rest/client/method/HttpSimplePostClientInvocation.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
+/*
```
[the remark this answers is not in the document]
Comment (volodymyr-korzh) [reply hapi-fhir-client/src/main/java/ca/uhn/fhir/rest/client/method/HttpSimplePostClientInvocation.java]: Great suggestion. Fixed

File: hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java (hunk 5 above)
Comment (volodymyr-korzh) [reply hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java]: You are right, removed

File: hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java (hunk 4 above)
Comment (volodymyr-korzh) [reply hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java]: Fixed.

File: hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java (hunk 3 above)
Comment (volodymyr-korzh) [reply hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java]: renamed

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/rest/gclient/IGetPageTyped.java [hunk 9]
```diff
@@ -20,8 +20,10 @@
 package ca.uhn.fhir.rest.gclient;
 
 
+import ca.uhn.fhir.rest.api.PageStyleEnum;
+
 public interface IGetPageTyped<T> extends IClientExecutable<IGetPageTyped<T>, T> {
-	
-	// nothing for now
+
```
[the remark this answers is not in the document]
Comment (volodymyr-korzh) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/rest/gclient/IGetPageTyped.java:26]: changed to usingMethod

File: hapi-fhir-client/src/main/java/ca/uhn/fhir/rest/client/impl/GenericClient.java [hunk 10]
```diff
@@ -848,20 +850,33 @@ private class GetPageInternal extends BaseClientExecutable<IGetPageTyped<Object>
...
 			myBundleType = theBundleType;
 		}
 
 		@Override
 		public Object execute() {
-			IClientResponseHandler binding;
-			binding = new ResourceResponseHandler(myBundleType, getPreferResponseTypes());
-			HttpSimpleGetClientInvocation invocation = new HttpSimpleGetClientInvocation(myContext, myUrl);
-			invocation.setUrlSource(UrlSourceEnum.EXPLICIT);
+			IClientResponseHandler binding = new ResourceResponseHandler(myBundleType, getPreferResponseTypes());
```
[the remark this answers is not in the document]
Comment (volodymyr-korzh) [reply hapi-fhir-client/src/main/java/ca/uhn/fhir/rest/client/impl/GenericClient.java:861]: Looks much clearer, fixed

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_8_0/5008-generic-client-post-paging-requests-support.yaml [hunk 11]
```diff
@@ -0,0 +1,5 @@
+---
+type: fix
+issue: 5008
+jira: SMILE-6727
```
[the remark this answers is not in the document]
Comment (volodymyr-korzh) [reply hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_8_0/5008-generic-client-post-paging-requests-support.yaml:4]: Fixed

File: hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java (hunk 6 above)
Comment (volodymyr-korzh) [reply hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java:316]: removed if/else. fixed

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/rest/api/PageStyleEnum.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (volodymyr-korzh) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/rest/api/PageStyleEnum.java:24]: maybe PageStyleEnum -> PagingRequestType ?

[the remark this answers is not in the document]
Comment (epeartree) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/rest/api/PageStyleEnum.java:24]: i'd settle for PagingHTTPMethodEnum.  <-- expresses what the enum is doing  without having to look at its component or read comments.  SearchStyleEnum is different.  searchStyleEnum is not use to make an http request but to make conditional branching  while processing a search request that which can be expressed 3 different ways.

**Review by epeartree — APPROVED** [review]

File: hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java [hunk 12]
```diff
@@ -307,18 +307,14 @@ public void setTransactionLatchEnabled(boolean theTransactionLatchEnabled) {
...
 		}
 
 		@Override
 		protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
 			// emulate a GET request invocation in the case of a POST paging request, as POST paging is not supported by FHIR RestServer
-			if (myExecutePostPagingRequestsEnabled) {
-				super.handleRequest(RequestTypeEnum.GET, request, response);
-			} else {
-				super.handleRequest(RequestTypeEnum.POST, request, response);
-			}
```
Comment (epeartree) [inline hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java]: `super.handleRequest(myPageStyle);`

Comment (volodymyr-korzh) [reply hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java]: Changed to `super.handleRequest(myPagingHttpMethod.getRequestType(), request, response);`. We can't reuse RequestTypeEnum names in PagingHttpMethodEnum but we can take these values in PagingHttpMethodEnum constructor and use them like a mapping.

File: hapi-fhir-client/src/main/java/ca/uhn/fhir/rest/client/method/HttpSimpleClientInvocation.java [hunk 13]
```diff
@@ -29,19 +30,22 @@
...
+
+	public HttpSimpleClientInvocation(FhirContext theContext, String theUrlPath, PageStyleEnum thePageStyleEnum) {
 		super(theContext);
 		myUrl = theUrlPath;
+		myPageStyleEnum = thePageStyleEnum;
 	}
 
 	@Override
 	public IHttpRequest asHttpRequest(String theUrlBase, Map<String, List<String>> theExtraParams, EncodingEnum theEncoding, Boolean thePrettyPrint) {
-		IHttpRequest retVal = createHttpRequest(myUrl, theEncoding, RequestTypeEnum.GET);
```
[the remark this answers is not in the document]
Comment (volodymyr-korzh) [reply hapi-fhir-client/src/main/java/ca/uhn/fhir/rest/client/method/HttpSimpleClientInvocation.java:44]: Changed to `IHttpRequest retVal = createHttpRequest(myUrl, theEncoding, myPagingHttpMethod.getRequestType());`. We can't reuse RequestTypeEnum names in PagingHttpMethodEnum but we can take these values in PagingHttpMethodEnum constructor and use them like a mapping.

File: hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java [hunk 14]
```diff
@@ -281,6 +286,8 @@ private static class MyRestfulServer extends RestfulServer {
 
 		private final boolean myInitialTransactionLatchEnabled;
 
+		private PagingHttpMethodEnum myPagingHttpMethod = PagingHttpMethodEnum.POST;
```
Comment (jmarchionatto) [inline hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java:289]: Isn't too much making POST the default? What is the reason?

**Review by jmarchionatto — APPROVED** [review]
Looks good! Approved with a question

File: hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java (hunk 14 above)
Comment (volodymyr-korzh) [reply hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/RestServerR4Helper.java]: You are right.
I have changed it to GET, as logic in RestServerR4Helpler may be confusing if we set `myPagingHttpMethod = PagingHttpMethodEnum.POST;`.
Later, for doPost method in RestServerR4Helpler we are checking what type of paging method we have, and emulating doGet, if paging method is POST.
`RequestTypeEnum requestType = myPagingHttpMethod == PagingHttpMethodEnum.POST ? RequestTypeEnum.GET : RequestTypeEnum.POST;`
`super.handleRequest(requestType, request, response);`

**Pull request merged**